pkg/kn/commands/revision/list.go

```diff
@@ -85,6 +86,19 @@ func NewRevisionListCommand(p *commands.KnParams) *cobra.Command {
 				}
 			}
 
+			// Add revision service info as annocations: traffic and tags
+			for _, revision := range revisionList.Items {
+				serviceName := revision.Labels[serving.ServiceLabelKey]
+				service, err := client.GetService(serviceName)
+				if err != nil {
+					return err
+				}
+
+				traffic, tags := trafficForRevision(revision.Name, service)
+				revision.Annotations[RevisionTrafficAnnotation] = fmt.Sprintf("%d%%", traffic)
+				revision.Annotations[RevisionTagsAnnotation] = strings.Join(tags, ",")
```

Comment:
Can we avoid modifying the revision here by e.g. using a  map keyed by revision name ? Imagine if someone later would introduce such annotations for real (so that you would override it here). Also 'misusing' an object by writing to it in a pure read-only use case is not architectural sound.

Reply:
Thought of that, however, the revision is not modified permanently. Only temporarily to hold values that frankly belongs to it. These modified revision objects are in memory and are never seen by 'anyone' else other than the user that invoked the command.

Reply:
> Can we avoid modifying the revision here by

+1

We can use `trafficForRevision` in `printRevision` to fetch the `traffic` and `tags`.

Reply:
Read my comment @navidshaikh it does not modify the revision except in memory which makes the rest of the code super clean

Reply:
Sure! IMO, this information doesn't belong to Revision and we shouldn't piggyback this data onto Revision.

Reply:
I'm still not sold to modify domain objects, even when only "in memory". You never know what happens with that objects. BTW, if you would do such thing with Hibernate (Java ORM mapper), it would be persisted at the end of a transaction, even when it's 'only in memory'. Also, I don't see why an object should be modified in a pure read-only situation (imagine that at some point in the future the Client API decides to make objects returned by read-only operations immutable by throwing an error on setting fields. Your code would break without a real reason.

It's just not a good style to misuse domain objects for temporary information. Not because something could be done easily should be done. (Imagine someone would print out the full object somewhere in between, like in a `kn revision list --verbose` or something. You will see artifical annotations here).

This can be achieved much more elegantly.

Reply:
I disagree since annotations are meant to do things like that. However, I am happy to hear suggestions on how else to do it?

Reply:
> Sure! I think this information doesn't belong to Revision and we shouldn't piggyback this data onto Revision.

It does at least when using the revision list. See why I opened this issue in the first place. It comes from the need to see this information and not having to have to show the details of the service and parsing the out to find the data. This is my feedback from real usage.

Reply:
> (Imagine someone would print out the full object somewhere in between, like in a `kn revision list --verbose` or something. You will see artifical annotations here).

`kn revision list --verbose` call does not exist. If we decide to do it we can easily filter out the annotations or better flag them as such. Which I'd be happy to also do here.

Reply:
> I disagree since annotations are meant to do things like that. However, I am happy to hear suggestions on how else to do it?

That's where I disagree, i.e. that annotations are used for volatile information which never gets persisted ;-). But maybe we could at least agree that listing revisions is a read-only operation and could be implemented with immutable objects that don't allow mutating operations? Why should we break that without a need ?

> `kn revision list --verbose` call does not exist. If we decide to do it we can easily filter out the annotations or better flag them as such. Which I'd be happy to also do here.

That's not really the point, whether this feature exists or not as its just an illustrative example. When you say one can easily add a filter, that is a bad smell, too, to use another hack to fix the previous hack. 

Not sure whether I could convince you that keeping certain architectural constraints even when they are not enforced today is a good thing. So let's merge them at let me make an alternative suggestion in the next PR.

Reply:
> Not sure whether I could convince you that keeping certain architectural constraints even when they are not enforced today is a good thing. 

That's not the part that I disagree with. 

> That's where I disagree, i.e. that annotations are used for volatile information which never gets persisted ;-). 

Yup, that's the conundrum. It's the use of annotations. For me, I feel that this is a good use since the attributes I am adding `traffic` and `tags` are not really "features" of the `Service` nor the `Revision` but of both. There are synthesized from and belong to the association between `Service` and `Revision`, in my view.

> So let's merge them at let me make an alternative suggestion in the next PR.

OK cool. Look forward to it. Hopefully it's not a significant change since that was where I was headed until this approach. But maybe I am missing some easy trick... Happy to learn.

Reply:
> OK cool. Look forward to it. Hopefully it's not a significant change since that was where I was headed until this approach. But maybe I am missing some easy trick... Happy to learn.

I have to apologize a bit :) Ideally I would use classical "view object" used for printing. But the printer framework we are using insist on `runtime.Object`. To implement those is even more hackish than adding to the annotations (I already was there before I returned back to your approach, which definitely is the better in this particular case). So I suggest indeed to stick to this technique for tunneling purposes (if we would have the printing of a list in our hands, its trivial to change with a  `revisionInfo` object containing the revision + tags,traffic from the service). 

However, my point is illustrated quite nicely what happens at the moment, when you are doing a `kn revision list -o yaml`. This includes, in YAML, the temporary annotations. This is incorrect as these annotations are *not part of the revision*. We can fix that, but in this initial PR it was overlooked. Something similar can happen in the future, too. Hopefully not, but anybody touching this code needs to be aware of this particular usage of volatile information attached to an otherwise persisted object.

So, there is no easy trick. Life is complicated ;-)